docs/Style.thy

```diff
@@ -0,0 +1,182 @@
+   indent statements that need to be broken over multiple lines. *)
+
+subsection \<open>Indentation (general)\<close>
+
+(* From Gerwin's guide:
+
+   If an apply command is applied to a state with n+1 subgoals, it must be indented by n spaces
+   relative to the first apply in the sequence (see the Isabelle distribution for many examples).
+   This indicates textually which line solved a subgoal and how many new subgoals where generated.
+   Invaluable information when something goes wrong because something changed somewhere. Also, it
+   gives a nice visual indication of the nesting depth of your proof.
+
+   The usual indentation for 0 subgoals is 2 spaces for a top-level proof.
+
+  For example:
+*)
+
+lemma test_lemma1:
+  "2 \<le> length l \<Longrightarrow> \<exists>h1 h2 t. l = h1 # h2 # t"
+  apply (case_tac l)
+   apply simp
+  apply (rename_tac h1 t1)
+  apply (case_tac t1)
+   apply simp
+  apply simp
+  done
+
+(* TO DO: Improve this example. *)
+
+subsection \<open>Indentation (multi-line proofs)\<close>
+
+(* If a line of proof will exceed 100 characters then it should be wrapped to the next line and
+   remain indented at a depth no less than it was already.
+
+   TO DO: Develop more rigorous advice on this issue.
+
+   Some examples.
+*)
+
+lemma test_lemma2:
+  "case_option None hd_opt (hd_opt l) = Some x \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                 split: list.splits)
+  done
+
+lemma test_lemma3:
+  "case_option None hd_opt (hd_opt l) = Some x \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply ((case_tac l; simp add: hd_opt_def),
+         rename_tac h t,
+         case_tac h; simp)
+  done
+
+subsection \<open>prefer and defer\<close>
+
+(* There are currently no hard rules on the use of `prefer` and `defer n`. Two general principles
+   apply:
+
+    1. If they are used too frequently then a proof may become unreadable. Do not use them
+       unnecessarily and consider including comments to indicate their purpose.
+    2. If they are used too "specifically" then a proof may break very frequently for not-
+       interesting reasons. This makes a proof less maintainable and so this should be avoided.
+
+   A known use case for `prefer` is where a proof has a tree-like structure, and the prover wants to
+   approach it with a breadth-first approach. Since the default isabelle strategy is depth-first,
+   prefers (or defers) will be needed, e.g. corres proofs.
+*)
+
+subsection \<open>auto\<close>
+
+(* The `auto` method must only be used at the end of a proof, as the last method in a proof.
+
+   TO DO: Elaborate on the reasons for this standard.
+*)
+
+subsection \<open>by\<close>
+
+(* The use of `by my_last_method` should be restricted to those cases where the final method
+   application is either:
+
+   1. Relatively simple and unlikely to break in the future, or
+   2. a very long-running method.
+
+   TO DO: Elaborate on the reasons for this standard.
```

Comment:
I thought we had actually decided that it doesn't matter in the end. The main complaint was that it's hard to figure out the error message if something breaks, but that has changed over Isabelle versions, and most methods do present their error message with `by`.

The reason I'd prefer not to make this is a rule is that upstream Isabelle style is to *prefer* `by`.

Reply:
Will remove for now.